### PR TITLE
Add faster Box2DWorld querying API

### PR DESCRIPTION
**Pull request by brianwinterpixel**
also resizes the output buffer on overflow rather than asserting false

### CODE REVIEW

**Comment by brianwinterpixel** [comment]
prob could have just committed to main

**Comment by brianwinterpixel** [comment]
nvm it's good i didn't

File: scene/2d/box2d_collision_object.cpp [hunk 1]
```diff
@@ -379,25 +379,25 @@ Array Box2DCollisionObject::get_colliding_bodies() const {
...
 			}
+
+			Box2DCollisionObject* godot_node = contact_iterator->other->GetUserData().owner;
+			p_array[i] = godot_node;
+			i++;
 		}
 		contact_iterator = contact_iterator->next;
 	}
+	if (p_array.size() > before_size) {
+		WARN_PRINT("Output buffer size was reallocated from " + itos(before_size) + " to " + itos(p_array.size()));
```
Comment (jasonwinterpixel) [inline scene/2d/box2d_collision_object.cpp:399]: Will this print() indicate that it originates from Box2DCollisionObject.get_colliding_bodies_fast ?

Comment (brianwinterpixel) [reply scene/2d/box2d_collision_object.cpp:399]: Yup, file and line number

File: scene/2d/box2d_world.cpp [hunk 2]
```diff
@@ -1373,6 +1378,95 @@ Array Box2DWorld::cast_motion(const Ref<Box2DShapeQueryParameters> &p_query) {
...
 
+int Box2DWorld::intersect_point_fast(Array p_out_array, const Vector2 &p_point, int p_max_results, const Array &p_exclude, uint32_t p_collision_mask, bool p_collide_with_bodies, bool p_collide_with_sensors, uint32_t p_collision_layer, int32_t p_group_index) {
+	// Memory-friendly impl that writes Box2DPhysicsBody outputs to a preallocated array
+	point_callback.obj_results.clear();
+	point_callback.point = gd_to_b2(p_point);
+	point_callback.max_results = p_max_results;
+	point_callback.report_obj_instead = true;
+
+	point_callback.exclude.clear();
+	for (int i = 0; i < p_exclude.size(); i++) {
```
Comment (jasonwinterpixel) [inline scene/2d/box2d_world.cpp:1389]: could probably support null for no excludes such that not even an empty array needs to be allocated

File: scene/2d/box2d_world.cpp [hunk 3]
```diff
@@ -712,6 +712,9 @@ void Box2DWorld::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("cast_motion", "query"), &Box2DWorld::cast_motion);
 	//ClassDB::bind_method(D_METHOD("collide_shape", "shape", "max_results"), &PhysicsDirectSpaceState2D::_collide_shape, DEFVAL(32));
 
+	ClassDB::bind_method(D_METHOD("intersect_point_fast", "output", "point", "max_results", "exclude", "collision_mask", "collide_with_bodies", "collide_with_sensors", "collision_layer", "group"), &Box2DWorld::intersect_point_fast, DEFVAL(32), DEFVAL(Array()), DEFVAL(0xFFFFFFFF), DEFVAL(true), DEFVAL(false), DEFVAL(0x0), DEFVAL(0));
```
Comment (jasonwinterpixel) [inline scene/2d/box2d_world.cpp:715]: Arent the default values 32 and Array() backwards? The default values dont seem to align with the parameters?

**Review by jasonwinterpixel — CHANGES_REQUESTED** [review]
Looks good, just curious about the default values looking like they dont line up to me

File: scene/2d/box2d_world.cpp (hunk 3 above)
Comment (brianwinterpixel) [reply scene/2d/box2d_world.cpp:715]: They look aligned to me. They align starting at the end of the parameter list

File: scene/2d/box2d_collision_object.cpp [hunk 4]
```diff
@@ -379,25 +379,25 @@ Array Box2DCollisionObject::get_colliding_bodies() const {
...
-				p_array[i] = godot_node;
-				i++;
-			}
-			else {
-				print_error("get_colliding_bodies overload");
-				assert(false);
+	int before_size = p_array.size();
+	while (contact_iterator) {	
+		if (contact_iterator->contact->IsTouching()) {
+			if (i >= p_array.size()) {
```
Comment (jasonwinterpixel) [inline scene/2d/box2d_collision_object.cpp]: this should be a while()

File: scene/2d/box2d_world.cpp (hunk 3 above)
Comment (jasonwinterpixel) [reply scene/2d/box2d_world.cpp:715]: I see

File: scene/2d/box2d_collision_object.cpp (hunk 4 above)
Comment (brianwinterpixel) [reply scene/2d/box2d_collision_object.cpp]: !

**Pull request merged**